enhancements/mcd-reboot-dictionary.md

```diff
@@ -0,0 +1,262 @@
+- Allowing different behaviour based on which option or part of the file has
+  been updated
+- Allowing reboots to be delayed so that several changes can be applied with 
+  a single reboot
+
+## Proposal
+
+The proposal calls for a whitelist that MCD uses to decide which action to take
+after updating a configuration file for the MCO.  To prevent the admin from
+tampering with the whitelist, it will be included as part of the MCD image as a
+static file in yaml format.
+
+The information could be baked into the the MCD itself, but assuming we treat it
+as data, the file will be a list of entries containing:
+
+- filename or glob
+- path
+- action to perform
+- drain required
+- action specific data
+
+Example (and not necessarily 100% accurate) entries:
+
+| File | Path | Action | Drain | Args |
+| ---- | ---- | ------ | ----- | ---- |
+| * | /var/home/core/.ssh/ | none | false ||
+| *.service | /etc/systemd/system/ | systemctl | false | _restart_ |
+| kubelet.conf | /etc/kubernetes/ | systemctl | false | _reload_ kubelet.service |
+| crio.conf    | /etc/crio/ | systemctl | true | _restart_ crio.service ||
+| imaginary.conf | /somewhere/ | binary | false | /bin/my-custom-tool -q |
+| grub2.conf | /etc/ | reboot | true ||
+
+After the MCD writes out a configuation change, it will consult the whitelist
+before deciding what action is required. If the filename is present, has a
+valid action and any associated data, the action will be performed.  Otherwise,
+a reboot will be performed as in prior 4.x versions.
+
+If a file contains any options that require a reboot to be applied, then any change 
+to it is treated as requiring a reboot.
+
+If a file matches more than one entry, the first one listed will be applied.
+
+If no service is supplied for a ``systemctl`` action, the matching filename (with
+directory components stripped) will be treated as the name of the unit file to be
+operated on.
+
+The system's highest priority is to apply the changes, if the specified action
+results in an error, then a reboot will be performed as in prior 4.x versions.
+
+Errors include 
+* the failure of any API calls (eg. https://github.com/coreos/go-systemd ) necessary 
+  to apply the change
+* external binaries/scripts (eg. `systemctl`) returning a value other than ``0``
+* actions taking "too" long (eg. we might want to put a limit on drain operations)
+
+In the case of multiple files being changed by a single update, if any file requires
+a reboot (either by configuration or an error), then the whitelist is ignored for all
+(remaining) files.  Otherwise, all files are handled independantly.
+
+
+
+### User Stories [optional]
+
+#### Story 1
+
+As an admin, I want to update the SSH keys without any followup action, to 
+reduce the number of unnecessary periods of degraded system and application 
+availability.
+
+#### Story 2
+
+As an admin, I want to update the CRI-O configuration by restarting it's systemd
```

Comment:
Can you enumerate the actual change desired?  It would help me understand what change is desired from crio that would also not want to drain and restart existing workload.

/cc @mrunalp

Reply:
I included a note in the Proposal section that might have been useful to repeat here in the Scenarios section too.  These stories are intended to be illustrative rather than authoritative.

In the specific case of CRI-O, earlier review conversations also pointed out that drain is necessary to apply changes to it's config file.  However the intention behind the story is to illustrate that there are also services (perhaps ``chrony`` would have been a better example) that can be fully applied with a restart of the corresponding systemd unit file.

The goal is to build the generic capability to avoid certain causes of reboots, rather than focus on a specific daemon or config file.  Whether cases that require drain fit into the "low-hanging fruit" category is less clear.